core/trino-main/src/main/java/io/trino/operator/InterpretedHashGenerator.java

```diff
@@ -45,26 +53,45 @@ public InterpretedHashGenerator(List<Type> hashChannelTypes, List<Integer> hashC
 
     public InterpretedHashGenerator(List<Type> hashChannelTypes, int[] hashChannels, BlockTypeOperators blockTypeOperators)
     {
-        this.hashChannels = requireNonNull(hashChannels, "hashChannels is null");
+        this(hashChannelTypes, requireNonNull(hashChannels, "hashChannels is null"), blockTypeOperators, false);
+    }
+
+    private InterpretedHashGenerator(List<Type> hashChannelTypes, @Nullable int[] hashChannels, BlockTypeOperators blockTypeOperators, boolean positional)
```

Comment:
Instead of a separate argument to indicate whether it's positional or not, why not signal it based on whether hashChannels is null or not?

Reply:
The constructor overload patterns were a little trickier to tease apart that way in terms of which validations require non-null but matching lengths vs which ones require null, etc. Adding a synthetic `boolean` argument made the overloads and validation logic easier to reason about when I was putting this together.

Reply:
Wouldn't something like this work? (I haven't checked it thoroughly, so I may be missing something)


```java
public static InterpretedHashGenerator createPositionalWithTypes(List<Type> hashChannelTypes, BlockTypeOperators blockTypeOperators)
{
    return new InterpretedHashGenerator(hashChannelTypes, (int[]) null, blockTypeOperators);
}

public InterpretedHashGenerator(List<Type> hashChannelTypes, List<Integer> hashChannels, BlockTypeOperators blockTypeOperators)
{
    this(hashChannelTypes, Ints.toArray(requireNonNull(hashChannels, "hashChannels is null")), blockTypeOperators);
}

private InterpretedHashGenerator(List<Type> hashChannelTypes, @Nullable int[] hashChannels, BlockTypeOperators blockTypeOperators)
{
    this.hashChannelTypes = ImmutableList.copyOf(requireNonNull(hashChannelTypes, "hashChannelTypes is null"));
    this.hashCodeOperators = createHashCodeOperators(hashChannelTypes, blockTypeOperators);

    checkArgument(hashCodeOperators.length == hashChannelTypes.size());
    checkArgument(hashChannels == null || hashChannels.length == hashCodeOperators.length);

    this.hashChannels = hashChannels;
}
```

Reply:
This overload: `List<Type> hashChannelTypes, @Nullable int[] hashChannels, BlockTypeOperators blockTypeOperators` is already public, but must _not_ pass null via that interface.

Reply:
Gotcha. That's what I overlooked.

In the future, we can improve this by moving to static factory methods for all cases instead of a mix of constructors and static methods.